File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 is None and list2 is None:
            return list1
        if list1 is None:
            return list2
        if list2 is None:
            return list1


        new = ListNode()
        new.next = None
        if list1.val > list2.val:
            new = list2
            list2=list2.next
        else:
            new = list1
            list1=list1.next

        placeholder = new
        

        while list1 != None and list2 !=None:
            print("list1.val: ", list1.val, "list2.val: ", list2.val)
            if list1.val > list2.val:
                new.next =list2
                list2=list2.next
            else:
                new.next = list1
                list1=list1.next
            new = new.next
        
        if list1 is None:
            new.next = list2
        else:
            new.next = list1
        
        return placeholder

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq; //val, id of the list
        for(int i=0; i <lists.size(); i++)
        {
            pq.push(make_pair(lists[i]->val, i));
        }
        bool is_root_col = false;
        ListNode* root = nullptr;
        ListNode* prev;
        while(!pq.empty())
        {
            auto [val, id]  = pq.top();
            pq.pop();
            ListNode* node = new ListNode(val);
            lists[id] = lists[id]->next;
            if(lists[id] != nullptr)
            {
                pq.push(make_pair(lists[id]->val, id));
            }
            
            if(!is_root_col)
            {
                root = node;
                is_root_col = true;
            }
            else{
                prev->next = node;
            }
            prev = node;
            
        }
        return root;
    }
};
